File: web/src/domain/backup.ts
export const BACKUP_FORMAT = 'fitlog-backup';
export const BACKUP_FORMAT_VERSION = 1;

export type BackupValue = string | number | null;
export type BackupRow = Record<string, BackupValue>;

export type TableName =
  | 'exercise'
  | 'routine'
  | 'routine_exercise'
  | 'session'
  | 'set_entry'
  | 'body_metric'
  | 'app_setting';

export const TABLES: readonly TableName[] = [
  'exercise',
  'routine',
  'routine_exercise',
  'session',
  'set_entry',
  'body_metric',
  'app_setting',
];

export interface BackupFile {
  readonly format: string;
  readonly format_version: number;
  readonly exported_at_ms: number;
  readonly app_version: string;
  readonly sections: Partial<Record<TableName, BackupRow[]>>;
}

export interface BackupSnapshot {
  exercise: BackupRow[];
  routine: BackupRow[];
  routine_exercise: BackupRow[];
  session: BackupRow[];
  set_entry: BackupRow[];
  body_metric: BackupRow[];
  app_setting: BackupRow[];
}

export interface SectionSummary {
  readonly inserted: number;
  readonly updated: number;
  readonly ignored: number;
  readonly remapped: number;
}

export type MergeSummary = Record<TableName, SectionSummary>;

export interface MergeResult {
  readonly snapshot: BackupSnapshot;
  readonly summary: MergeSummary;
}

export type BackupErrorCode =
  | 'invalid_json'
  | 'unknown_format'
  | 'unsupported_version'
  | 'invalid_row'
  | 'missing_reference';

export class BackupError extends Error {
  readonly code: BackupErrorCode;
  readonly detail: string;

  constructor(code: BackupErrorCode, detail: string) {
    super(`${code}: ${detail}`);
    this.name = 'BackupError';
    this.code = code;
    this.detail = detail;
  }
}

interface TableSpec {
  readonly name: TableName;
  readonly pk: string;
  readonly required: readonly string[];
}

const TABLE_SPECS: Record<TableName, TableSpec> = {
  exercise: {
    name: 'exercise',
    pk: 'id',
    required: ['id', 'slug', 'name', 'muscle_group_id', 'is_custom', 'updated_at'],
  },
  routine: { name: 'routine', pk: 'id', required: ['id', 'name', 'updated_at'] },
  routine_exercise: {
    name: 'routine_exercise',
    pk: 'id',
    required: ['id', 'routine_id', 'exercise_id', 'position', 'updated_at'],
  },
  session: { name: 'session', pk: 'id', required: ['id', 'started_at', 'updated_at'] },
  set_entry: {
    name: 'set_entry',
    pk: 'id',
    required: ['id', 'session_id', 'exercise_id', 'set_index', 'updated_at'],
  },
  body_metric: {
    name: 'body_metric',
    pk: 'id',
    required: ['id', 'measured_at', 'kind', 'value', 'unit', 'updated_at'],
  },
  app_setting: { name: 'app_setting', pk: 'key', required: ['key', 'value', 'updated_at'] },
};

interface ReferenceSpec {
  readonly table: TableName;
  readonly column: string;
  readonly target: TableName;
}

const REFERENCES: readonly ReferenceSpec[] = [
  { table: 'exercise', column: 'muscle_group_id', target: 'exercise' },
  { table: 'routine_exercise', column: 'routine_id', target: 'routine' },
  { table: 'routine_exercise', column: 'exercise_id', target: 'exercise' },
  { table: 'set_entry', column: 'session_id', target: 'session' },
  { table: 'set_entry', column: 'exercise_id', target: 'exercise' },
];

export function emptySnapshot(): BackupSnapshot {
  return {
    exercise: [],
    routine: [],
    routine_exercise: [],
    session: [],
    set_entry: [],
    body_metric: [],
    app_setting: [],
  };
}

export function parseBackup(json: string): BackupFile {
  let parsed: unknown;
  try {
    parsed = JSON.parse(json);
  } catch {
    throw new BackupError('invalid_json', 'El archivo no es JSON válido');
  }

  if (typeof parsed !== 'object' || parsed === null) {
    throw new BackupError('invalid_json', 'El archivo no contiene un objeto JSON');
  }

  const candidate = parsed as Partial<BackupFile>;
  if (candidate.format !== BACKUP_FORMAT) {
    throw new BackupError(
      'unknown_format',
      `Se esperaba el formato ${BACKUP_FORMAT} y llegó ${String(candidate.format)}`
    );
  }
  if (typeof candidate.format_version !== 'number') {
    throw new BackupError('invalid_json', 'Falta format_version');
  }
  if (candidate.format_version > BACKUP_FORMAT_VERSION) {
    throw new BackupError(
      'unsupported_version',
      `El archivo es de la versión ${candidate.format_version} y esta app soporta hasta la ${BACKUP_FORMAT_VERSION}`
    );
  }

  const sections: Partial<Record<TableName, BackupRow[]>> = {};
  for (const table of TABLES) {
    const rows = candidate.sections?.[table];
    if (rows === undefined) continue;
    if (!Array.isArray(rows)) {
      throw new BackupError('invalid_json', `La sección ${table} no es una lista`);
    }
    sections[table] = rows as BackupRow[];
  }

  return {
    format: candidate.format,
    format_version: candidate.format_version,
    exported_at_ms: typeof candidate.exported_at_ms === 'number' ? candidate.exported_at_ms : 0,
    app_version: typeof candidate.app_version === 'string' ? candidate.app_version : '',
    sections,
  };
}

function requiredFields(table: TableName): readonly string[] {
  return TABLE_SPECS[table].required;
}

function primaryKey(table: TableName): string {
  return TABLE_SPECS[table].pk;
}

function isPresent(row: BackupRow, field: string): boolean {
  return Object.prototype.hasOwnProperty.call(row, field) && row[field] !== undefined;
}

export function validateBackup(file: BackupFile, local: BackupSnapshot): void {
  for (const table of TABLES) {
    const rows = file.sections[table];
    if (!rows) continue;
    for (const row of rows) {
      for (const field of requiredFields(table)) {
        if (!isPresent(row, field)) {
          const identifier = String(row[primaryKey(table)] ?? '?');
          throw new BackupError(
            'invalid_row',
            `La fila ${identifier} de ${table} no tiene el campo ${field}`
          );
        }
      }
    }
  }

  const known = new Map<string, Set<string>>();
  for (const table of TABLES) {
    const ids = new Set<string>();
    for (const row of local[table]) {
      ids.add(String(row[primaryKey(table)]));
    }
    for (const row of file.sections[table] ?? []) {
      ids.add(String(row[primaryKey(table)]));
    }
    known.set(table, ids);
  }

  const localGroups = new Set(local.exercise.map((row) => String(row.muscle_group_id)));
  const fileGroups = new Set(
    (file.sections.exercise ?? []).map((row) => String(row.muscle_group_id))
  );

  for (const reference of REFERENCES) {
    const rows = file.sections[reference.table] ?? [];
    for (const row of rows) {
      const value = row[reference.column];
      if (value === null || value === undefined) continue;
      const key = String(value);

      if (reference.table === 'exercise' && reference.column === 'muscle_group_id') {
        if (!localGroups.has(key) && !fileGroups.has(key)) {
          throw new BackupError(
            'missing_reference',
            `El ejercicio ${String(row.id)} referencia el grupo muscular ${key}, que no existe en la base`
          );
        }
        continue;
      }

      const candidates = known.get(reference.target);
      if (!candidates?.has(key)) {
        throw new BackupError(
          'missing_reference',
          `La fila ${String(row[primaryKey(reference.table)])} de ${reference.table} referencia ${reference.target} ${key}, que no está en el archivo ni en la base`
        );
      }
    }
  }
}

function updatedAt(row: BackupRow): number {
  const value = row.updated_at;
  return typeof value === 'number' ? value : 0;
}

function mergeByPrimaryKey(
  localRows: readonly BackupRow[],
  importedRows: readonly BackupRow[],
  pk: string
): { rows: BackupRow[]; inserted: number; updated: number; ignored: number } {
  const byId = new Map<string, BackupRow>();
  for (const row of localRows) {
    byId.set(String(row[pk]), row);
  }

  let inserted = 0;
  let updated = 0;
  let ignored = 0;

  for (const imported of importedRows) {
    const key = String(imported[pk]);
    const existing = byId.get(key);
    if (!existing) {
      byId.set(key, imported);
      inserted += 1;
      continue;
    }
    if (updatedAt(imported) > updatedAt(existing)) {
      byId.set(key, imported);
      updated += 1;
    } else {
      ignored += 1;
    }
  }

  const rows = [...byId.values()].sort((a, b) => {
    const keyA = String(a[pk]);
    const keyB = String(b[pk]);
    return keyA < keyB ? -1 : keyA > keyB ? 1 : 0;
  });
  return { rows, inserted, updated, ignored };
}

function remapRow(row: BackupRow, column: string, idMap: Map<string, string>): BackupRow {
  const value = row[column];
  if (value === null || value === undefined) return row;
  const mapped = idMap.get(String(value));
  if (mapped === undefined) return row;
  return { ...row, [column]: mapped };
}

export function mergeBackup(local: BackupSnapshot, file: BackupFile): MergeResult {
  validateBackup(file, local);

  const snapshot: BackupSnapshot = {
    exercise: [...local.exercise],
    routine: [...local.routine],
    routine_exercise: [...local.routine_exercise],
    session: [...local.session],
    set_entry: [...local.set_entry],
    body_metric: [...local.body_metric],
    app_setting: [...local.app_setting],
  };
  const summary = {} as MergeSummary;
  for (const table of TABLES) {
    summary[table] = { inserted: 0, updated: 0, ignored: 0, remapped: 0 };
  }

  const idMap = new Map<string, string>();
  const localExerciseById = new Map(snapshot.exercise.map((row) => [String(row.id), row]));
  const localExerciseBySlug = new Map(snapshot.exercise.map((row) => [String(row.slug), row]));

  const importedExercises: BackupRow[] = [];
  for (const imported of file.sections.exercise ?? []) {
    const id = String(imported.id);
    const existingById = localExerciseById.get(id);
    if (existingById) {
      importedExercises.push(imported);
      continue;
    }
    const existingBySlug = localExerciseBySlug.get(String(imported.slug));
    if (existingBySlug) {
      idMap.set(id, String(existingBySlug.id));
      summary.exercise = { ...summary.exercise, remapped: summary.exercise.remapped + 1 };
      continue;
    }
    importedExercises.push(imported);
  }

  const exerciseMerge = mergeByPrimaryKey(
    snapshot.exercise,
    importedExercises,
    primaryKey('exercise')
  );
  snapshot.exercise = exerciseMerge.rows;
  summary.exercise = {
    ...summary.exercise,
    inserted: exerciseMerge.inserted,
    updated: exerciseMerge.updated,
    ignored: exerciseMerge.ignored,
  };

  for (const table of TABLES) {
    if (table === 'exercise') continue;
    const importedRows = (file.sections[table] ?? []).map((row) => {
      if (table === 'routine_exercise' || table === 'set_entry') {
        return remapRow(row, 'exercise_id', idMap);
      }
      return row;
    });

    const merged = mergeByPrimaryKey(snapshot[table], importedRows, primaryKey(table));
    snapshot[table] = merged.rows;
    summary[table] = {
      inserted: merged.inserted,
      updated: merged.updated,
      ignored: merged.ignored,
      remapped: 0,
    };
  }

  return { snapshot, summary };
}

export interface ExportOptions {
  readonly sections: readonly TableName[];
  readonly appVersion: string;
  readonly nowMs: number;
}

export function serializeBackup(
  snapshot: BackupSnapshot,
  options: ExportOptions
): string {
  const requested = new Set<TableName>(options.sections);

  const sets = requested.has('set_entry') ? snapshot.set_entry : [];
  const routineItems = requested.has('routine_exercise') ? snapshot.routine_exercise : [];

  const referencedExerciseIds = new Set<string>();
  for (const set of sets) {
    referencedExerciseIds.add(String(set.exercise_id));
  }
  for (const item of routineItems) {
    referencedExerciseIds.add(String(item.exercise_id));
  }

  const exercises = snapshot.exercise.filter(
    (row) =>
      Number(row.is_custom) === 1 &&
      (requested.has('exercise') || referencedExerciseIds.has(String(row.id)))
  );

  const sections: Partial<Record<TableName, BackupRow[]>> = {};
  if (exercises.length > 0) sections.exercise = exercises;
  if (requested.has('routine')) sections.routine = snapshot.routine;
  if (routineItems.length > 0) sections.routine_exercise = routineItems;
  if (requested.has('session')) sections.session = snapshot.session;
  if (sets.length > 0) sections.set_entry = sets;
  if (requested.has('body_metric')) sections.body_metric = snapshot.body_metric;
  if (requested.has('app_setting')) sections.app_setting = snapshot.app_setting;

  const file: BackupFile = {
    format: BACKUP_FORMAT,
    format_version: BACKUP_FORMAT_VERSION,
    exported_at_ms: options.nowMs,
    app_version: options.appVersion,
    sections,
  };

  return JSON.stringify(file, null, 2);
}

export const UI_GROUPS: readonly { readonly id: string; readonly label: string; readonly tables: readonly TableName[] }[] = [
  { id: 'custom_exercises', label: 'Ejercicios propios', tables: ['exercise'] },
  { id: 'routines', label: 'Rutinas', tables: ['routine', 'routine_exercise'] },
  { id: 'workouts', label: 'Entrenamientos', tables: ['session', 'set_entry'] },
  { id: 'body_metrics', label: 'Medidas', tables: ['body_metric'] },
  { id: 'settings', label: 'Ajustes', tables: ['app_setting'] },
];

File: android/app/src/main/java/com/fitlog/app/domain/Backup.kt
package com.fitlog.app.domain

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonPrimitive

object Backup {

    const val FORMAT = "fitlog-backup"
    const val FORMAT_VERSION = 1

    val TABLES = listOf(
        "exercise",
        "routine",
        "routine_exercise",
        "session",
        "set_entry",
        "body_metric",
        "app_setting",
    )

    enum class ErrorCode { INVALID_JSON, UNKNOWN_FORMAT, UNSUPPORTED_VERSION, INVALID_ROW, MISSING_REFERENCE }

    class BackupException(
        val code: ErrorCode,
        val detail: String,
    ) : Exception("$code: $detail")

    data class Row(val values: Map<String, Any?>)

    data class File(
        val format: String,
        val formatVersion: Int,
        val exportedAtMs: Long,
        val appVersion: String,
        val sections: Map<String, List<Row>>,
    )

    data class Snapshot(val tables: MutableMap<String, MutableList<Row>>) {
        fun rows(table: String): MutableList<Row> = tables.getOrPut(table) { mutableListOf() }
    }

    data class SectionSummary(
        val inserted: Int = 0,
        val updated: Int = 0,
        val ignored: Int = 0,
        val remapped: Int = 0,
    )

    data class MergeResult(
        val snapshot: Snapshot,
        val summary: Map<String, SectionSummary>,
    )

    private data class TableSpec(val pk: String, val required: List<String>)

    private val TABLE_SPECS = mapOf(
        "exercise" to TableSpec(
            pk = "id",
            required = listOf("id", "slug", "name", "muscle_group_id", "is_custom", "updated_at"),
        ),
        "routine" to TableSpec(pk = "id", required = listOf("id", "name", "updated_at")),
        "routine_exercise" to TableSpec(
            pk = "id",
            required = listOf("id", "routine_id", "exercise_id", "position", "updated_at"),
        ),
        "session" to TableSpec(pk = "id", required = listOf("id", "started_at", "updated_at")),
        "set_entry" to TableSpec(
            pk = "id",
            required = listOf("id", "session_id", "exercise_id", "set_index", "updated_at"),
        ),
        "body_metric" to TableSpec(
            pk = "id",
            required = listOf("id", "measured_at", "kind", "value", "unit", "updated_at"),
        ),
        "app_setting" to TableSpec(pk = "key", required = listOf("key", "value", "updated_at")),
    )

    private data class Reference(val table: String, val column: String, val target: String)

    private val REFERENCES = listOf(
        Reference("exercise", "muscle_group_id", "exercise"),
        Reference("routine_exercise", "routine_id", "routine"),
        Reference("routine_exercise", "exercise_id", "exercise"),
        Reference("set_entry", "session_id", "session"),
        Reference("set_entry", "exercise_id", "exercise"),
    )

    fun emptySnapshot(): Snapshot = Snapshot(TABLES.associateWith { mutableListOf() }.toMutableMap())

    private fun JsonElement.toValue(): Any? = when (this) {
        is JsonNull -> null
        is JsonPrimitive -> if (isString) content else doubleOrNull ?: content.toBooleanStrictOrNull() ?: content
        is JsonArray -> throw BackupException(ErrorCode.INVALID_JSON, "No se admiten listas dentro de una fila")
        is JsonObject -> throw BackupException(ErrorCode.INVALID_JSON, "No se admiten objetos dentro de una fila")
    }

    fun parse(json: String): File {
        val root = try {
            Json.parseToJsonElement(json)
        } catch (error: Exception) {
            throw BackupException(ErrorCode.INVALID_JSON, "El archivo no es JSON válido")
        }
        if (root !is JsonObject) {
            throw BackupException(ErrorCode.INVALID_JSON, "El archivo no contiene un objeto JSON")
        }

        val format = root["format"]?.jsonPrimitive?.content
        if (format != FORMAT) {
            throw BackupException(
                ErrorCode.UNKNOWN_FORMAT,
                "Se esperaba el formato $FORMAT y llegó ${format ?: "sin formato"}",
            )
        }

        val version = root["format_version"]?.jsonPrimitive?.doubleOrNull?.toInt()
            ?: throw BackupException(ErrorCode.INVALID_JSON, "Falta format_version")
        if (version > FORMAT_VERSION) {
            throw BackupException(
                ErrorCode.UNSUPPORTED_VERSION,
                "El archivo es de la versión $version y esta app soporta hasta la $FORMAT_VERSION",
            )
        }

        val sections = mutableMapOf<String, List<Row>>()
        val sectionsObject = root["sections"] as? JsonObject
        if (sectionsObject != null) {
            for (table in TABLES) {
                val array = sectionsObject[table] as? JsonArray ?: continue
                sections[table] = array.map { element ->
                    val rowObject = element as? JsonObject
                        ?: throw BackupException(ErrorCode.INVALID_JSON, "La sección $table no contiene objetos")
                    Row(rowObject.mapValues { (_, value) -> value.toValue() })
                }
            }
        }

        return File(
            format = format,
            formatVersion = version,
            exportedAtMs = root["exported_at_ms"]?.jsonPrimitive?.doubleOrNull?.toLong() ?: 0L,
            appVersion = root["app_version"]?.jsonPrimitive?.content ?: "",
            sections = sections,
        )
    }

    private fun required(table: String): List<String> = TABLE_SPECS.getValue(table).required

    private fun primaryKey(table: String): String = TABLE_SPECS.getValue(table).pk

    fun validate(file: File, local: Snapshot) {
        for (table in TABLES) {
            val rows = file.sections[table] ?: continue
            for (row in rows) {
                for (field in required(table)) {
                    if (!row.values.containsKey(field)) {
                        val identifier = row.values[primaryKey(table)]?.toString() ?: "?"
                        throw BackupException(
                            ErrorCode.INVALID_ROW,
                            "La fila $identifier de $table no tiene el campo $field",
                        )
                    }
                }
            }
        }

        val known = mutableMapOf<String, MutableSet<String>>()
        for (table in TABLES) {
            val ids = mutableSetOf<String>()
            local.rows(table).forEach { row -> ids.add(row.values[primaryKey(table)]?.toString() ?: "") }
            (file.sections[table] ?: emptyList()).forEach { row ->
                ids.add(row.values[primaryKey(table)]?.toString() ?: "")
            }
            known[table] = ids
        }

        val localGroups = local.rows("exercise").mapNotNull { it.values["muscle_group_id"]?.toString() }.toSet()
        val fileGroups = (file.sections["exercise"] ?: emptyList())
            .mapNotNull { it.values["muscle_group_id"]?.toString() }
            .toSet()

        for (reference in REFERENCES) {
            val rows = file.sections[reference.table] ?: continue
            for (row in rows) {
                val value = row.values[reference.column]?.toString() ?: continue
                if (reference.table == "exercise" && reference.column == "muscle_group_id") {
                    if (value !in localGroups && value !in fileGroups) {
                        throw BackupException(
                            ErrorCode.MISSING_REFERENCE,
                            "El ejercicio ${row.values["id"]} referencia el grupo muscular $value, que no existe en la base",
                        )
                    }
                    continue
                }
                val candidates = known[reference.target]
                if (candidates == null || value !in candidates) {
                    throw BackupException(
                        ErrorCode.MISSING_REFERENCE,
                        "La fila ${row.values[primaryKey(reference.table)]} de ${reference.table} " +
                            "referencia ${reference.target} $value, que no está en el archivo ni en la base",
                    )
                }
            }
        }
    }

    private fun updatedAt(row: Row): Double = (row.values["updated_at"] as? Double) ?: 0.0

    private data class MergeCounts(val rows: MutableList<Row>, val inserted: Int, val updated: Int, val ignored: Int)

    private fun mergeByPrimaryKey(
        localRows: List<Row>,
        importedRows: List<Row>,
        pk: String,
    ): MergeCounts {
        val byId = linkedMapOf<String, Row>()
        localRows.forEach { row -> byId[row.values[pk]?.toString() ?: ""] = row }

        var inserted = 0
        var updated = 0
        var ignored = 0

        for (imported in importedRows) {
            val key = imported.values[pk]?.toString() ?: ""
            val existing = byId[key]
            when {
                existing == null -> {
                    byId[key] = imported
                    inserted += 1
                }
                updatedAt(imported) > updatedAt(existing) -> {
                    byId[key] = imported
                    updated += 1
                }
                else -> ignored += 1
            }
        }

        val rows = byId.entries
            .sortedBy { it.key }
            .map { it.value }
            .toMutableList()
        return MergeCounts(rows, inserted, updated, ignored)
    }

    private fun remap(row: Row, column: String, idMap: Map<String, String>): Row {
        val value = row.values[column]?.toString() ?: return row
        val mapped = idMap[value] ?: return row
        return Row(row.values + (column to mapped))
    }

    fun merge(local: Snapshot, file: File): MergeResult {
        validate(file, local)

        val snapshot = Snapshot(
            TABLES.associateWith { table -> local.rows(table).toMutableList() }.toMutableMap()
        )
        val summary = TABLES.associateWith { SectionSummary() }.toMutableMap()

        val idMap = mutableMapOf<String, String>()
        val localById = snapshot.rows("exercise").associateBy { it.values["id"]?.toString() ?: "" }
        val localBySlug = snapshot.rows("exercise").associateBy { it.values["slug"]?.toString() ?: "" }

        val importedExercises = mutableListOf<Row>()
        for (imported in file.sections["exercise"] ?: emptyList()) {
            val id = imported.values["id"]?.toString() ?: ""
            if (localById.containsKey(id)) {
                importedExercises.add(imported)
                continue
            }
            val existingBySlug = localBySlug[imported.values["slug"]?.toString() ?: ""]
            if (existingBySlug != null) {
                idMap[id] = existingBySlug.values["id"]?.toString() ?: ""
                summary["exercise"] = summary.getValue("exercise").copy(
                    remapped = summary.getValue("exercise").remapped + 1,
                )
                continue
            }
            importedExercises.add(imported)
        }

        val exerciseMerge = mergeByPrimaryKey(snapshot.rows("exercise"), importedExercises, primaryKey("exercise"))
        snapshot.tables["exercise"] = exerciseMerge.rows
        summary["exercise"] = summary.getValue("exercise").copy(
            inserted = exerciseMerge.inserted,
            updated = exerciseMerge.updated,
            ignored = exerciseMerge.ignored,
        )

        for (table in TABLES) {
            if (table == "exercise") continue
            val importedRows = (file.sections[table] ?: emptyList()).map { row ->
                if (table == "routine_exercise" || table == "set_entry") {
                    remap(row, "exercise_id", idMap)
                } else {
                    row
                }
            }
            val merged = mergeByPrimaryKey(snapshot.rows(table), importedRows, primaryKey(table))
            snapshot.tables[table] = merged.rows
            summary[table] = SectionSummary(
                inserted = merged.inserted,
                updated = merged.updated,
                ignored = merged.ignored,
            )
        }

        return MergeResult(snapshot = snapshot, summary = summary)
    }

    fun serialize(
        snapshot: Snapshot,
        sections: List<String>,
        appVersion: String,
        nowMs: Long,
    ): String {
        val requested = sections.toSet()

        val sets = if ("set_entry" in requested) snapshot.rows("set_entry") else emptyList()
        val routineItems = if ("routine_exercise" in requested) snapshot.rows("routine_exercise") else emptyList()

        val referencedExerciseIds = mutableSetOf<String>()
        sets.forEach { row -> row.values["exercise_id"]?.toString()?.let { referencedExerciseIds.add(it) } }
        routineItems.forEach { row ->
            row.values["exercise_id"]?.toString()?.let { referencedExerciseIds.add(it) }
        }

        val exercises = snapshot.rows("exercise").filter { row ->
            val isCustom = (row.values["is_custom"] as? Double)?.toInt() == 1
            val id = row.values["id"]?.toString() ?: ""
            isCustom && ("exercise" in requested || id in referencedExerciseIds)
        }

        val selected = linkedMapOf<String, List<Row>>()
        if (exercises.isNotEmpty()) selected["exercise"] = exercises
        if ("routine" in requested) selected["routine"] = snapshot.rows("routine")
        if (routineItems.isNotEmpty()) selected["routine_exercise"] = routineItems
        if ("session" in requested) selected["session"] = snapshot.rows("session")
        if (sets.isNotEmpty()) selected["set_entry"] = sets
        if ("body_metric" in requested) selected["body_metric"] = snapshot.rows("body_metric")
        if ("app_setting" in requested) selected["app_setting"] = snapshot.rows("app_setting")

        val root = buildJsonObject {
            put("format", JsonPrimitive(FORMAT))
            put("format_version", JsonPrimitive(FORMAT_VERSION))
            put("exported_at_ms", JsonPrimitive(nowMs))
            put("app_version", JsonPrimitive(appVersion))
            put(
                "sections",
                buildJsonObject {
                    for ((table, rows) in selected) {
                        put(
                            table,
                            buildJsonArray {
                                rows.forEach { row ->
                                    add(
                                        buildJsonObject {
                                            for ((column, value) in row.values) {
                                                when (value) {
                                                    null -> put(column, JsonNull)
                                                    is String -> put(column, JsonPrimitive(value))
                                                    is Double -> put(
                                                        column,
                                                        JsonPrimitive(
                                                            if (value == value.toLong().toDouble()) value.toLong() else value
                                                        ),
                                                    )
                                                    else -> put(column, JsonPrimitive(value.toString()))
                                                }
                                            }
                                        }
                                    )
                                }
                            },
                        )
                    }
                },
            )
        }

        return Json { prettyPrint = true }.encodeToString(JsonObject.serializer(), root)
    }
}
